### PR TITLE
ramips: add support for Netgear AC 1200 R6120

### PR DESCRIPTION
**Pull request by Brother-Lal**
Hello!
These commits should add support for the Netgear dual-band router r6120, as well as basic support for the Sercomm/Netgear firmware file format.

It is not 100% in a mergeable state, but i wanted to get feedback on some issues.
(and ofc, all other issues that are still present... :+1: )

Open Questions:
1. The GPIO/LEDs are all working, but the ralink,group is most likely not correct - somebody with more info on the mappings should clarify that.
2. The firmware format is accepted by the nmrp as well as the webinterface, but the webinterface only copies a first part. Anybody has any ideas on that - i am willing to provide flashdumps and other support if needed, as i am out of ideas.

Otherwise I tested all things i know, macs, wifi, ethernet, sysupgrade is all working as it should.

Greetings,

Brother Lal

### CODE REVIEW

File: target/linux/ramips/base-files/etc/board.d/01_leds [hunk 1]
```diff
@@ -298,6 +298,12 @@ mzk-ex750np)
 na930)
 	set_usb_led "$boardname:blue:status"
 	;;
+netgear,r6120)
+	ucidef_set_led_default "power" "power" "$boardname:green:power" "1"
```
Comment (mkresin) [inline target/linux/ramips/base-files/etc/board.d/01_leds]: What is the purpose of this line. Let diag.sh handle the led.

File: target/linux/ramips/base-files/etc/board.d/01_leds [hunk 2]
```diff
@@ -298,6 +298,12 @@ mzk-ex750np)
...
+	ucidef_set_led_netdev "lan" "lan" "$boardname:green:lan" "eth0.2"
```
Comment (mkresin) [inline target/linux/ramips/base-files/etc/board.d/01_leds]: depending on the configuration, the lan ports might not be eth0.2. Please use `ucidef_set_led_switch` instead. the last argument is a bitmask of ports to trigger.

But if eth0.2 is your lan port something is wrong. According to your configuration, eth0.1 should be lan.

File: target/linux/ramips/dts/R6120.dts [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+		#address-cells = <1>;
+		#size-cells = <0>;
+		poll-interval = <20>;
+
+		reset {
+			label = "reset";
+			gpios = <&gpio1 6 GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_RESTART>;
+		};
+
```
Comment (mkresin) [inline target/linux/ramips/dts/R6120.dts]: drop this blank line

File: target/linux/ramips/dts/R6120.dts [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+			label = "r6120:green:power";
+			gpios = <&gpio1 11 GPIO_ACTIVE_LOW>;
+		};
+
+		wlan {
+			label = "r6120:green:wlan2g";
+			gpios = <&gpio1 10 GPIO_ACTIVE_LOW>;
+		};
+
+                wlan_orange {
```
Comment (mkresin) [inline target/linux/ramips/dts/R6120.dts]: spaces vs. tabs here. please check the whole dts for similar issues.

File: target/linux/ramips/dts/R6120.dts [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+			read-only;
+		};
+
+		factory: partition@40000 {
+			label = "factory";
+			reg = <0x40000 0x20000>;
+			read-only;
+		};
+
+		nvram: partition@60000 {
```
Comment (mkresin) [inline target/linux/ramips/dts/R6120.dts]: any reason why this partition isn't marked as read-only as well

File: target/linux/ramips/dts/R6120.dts [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+			label = "firmware";
+			reg = <0x90000 0xf60000>;
+		};
+
+		partition@ff0000 {
+			label = "reserved";
+			reg = <0xff0000 0x10000>;
+			read-only;
+		};
+
```
Comment (mkresin) [inline target/linux/ramips/dts/R6120.dts]: drop these and the following blank line

File: target/linux/ramips/dts/R6120.dts [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+	status = "okay";
+
+	pcie-bridge {
+		mt76@0,0 {
+			reg = <0x0000 0 0 0 0>;
+			device_type = "pci";
+			mediatek,mtd-eeprom = <&factory 0x28000>;
+			ieee80211-freq-limit = <5000000 6000000>;
+			mtd-mac-address = <&nvram 0x100B0>;
+			mtd-mac-address-increment = <(2)>;
```
Comment (mkresin) [inline target/linux/ramips/dts/R6120.dts]: Where is the base mac address incremented by one used?

File: target/linux/ramips/dts/R6120.dts [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+		};
+	};
+};
+
```
Comment (mkresin) [inline target/linux/ramips/dts/R6120.dts]: blank lines

File: target/linux/ramips/image/mt76x8.mk [hunk 9]
```diff
@@ -90,6 +90,23 @@ define Device/mt7628
 endef
 TARGET_DEVICES += mt7628
 
+define Device/netgear_r6120
+  DTS := R6120
+  BLOCKSIZE := 64k
+  IMAGE_SIZE := $(ralink_default_fw_size_8M)
```
Comment (mkresin) [inline target/linux/ramips/image/mt76x8.mk]: ralink_default_fw_size_16M

File: target/linux/ramips/image/mt76x8.mk [hunk 10]
```diff
@@ -90,6 +90,23 @@ define Device/mt7628
...
+  DEVICE_TITLE := Netgear AC1200 R6120
+  DEVICE_PACKAGES := kmod-usb2 kmod-usb-ohci kmod-usb-ledtrig-usbport
```
Comment (mkresin) [inline target/linux/ramips/image/mt76x8.mk]: no usb led, no kmod-usb-ledtrig-usbport

File: target/linux/ramips/dts/R6120.dts [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+                        gpios = <&gpio1 7 GPIO_ACTIVE_LOW>;
+                };
+
+	};
+};
+
+&pinctrl {
+	state_default: pinctrl0 {
+		gpio {
+			ralink,group = "i2c", "gpio", "p0led_an", "p1led_an", "p2led_an",
```
Comment (mkresin) [inline target/linux/ramips/dts/R6120.dts]: drop the i2c and gpio pins from the group. none of the pins is used
bildschirmfoto vom 2018-06-21 23-48-03

File: target/linux/ramips/dts/R6120.dts [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+		wlan5 {
+			label = "r6120:green:wlan5g";
+			gpios = <&gpio1 8 GPIO_ACTIVE_LOW>;
+		};
+
+                wlan5_orange {
+                        label = "r6120:orange:wlan5g";
+                        gpios = <&gpio1 7 GPIO_ACTIVE_LOW>;
+                };
+
```
Comment (mkresin) [inline target/linux/ramips/dts/R6120.dts]: drop this blank line

**Review by mkresin — CHANGES_REQUESTED** [review]
I've only done a quick review. Expect more changes if I've the time do do a full review.

Please use ```git commit --amend``` to add changes to your commit. Use ```git push --force``` to overwrite your github repository/branch. The PR will afterwards update automatically. Please don't close this PR and open a new one.

**Comment by Brother-Lal** [comment]
Thanks for the review, i'll try to do the request changes soonish.
For the gpio-mappings, is that sheet somewhere available? I had a link at some point to a different png of the mappings, but lost it.
Would be good for the next time to check.

File: target/linux/ramips/dts/R6120.dts (hunk 5 above)
Comment (Brother-Lal) [reply target/linux/ramips/dts/R6120.dts]: I though i needed to modify nvram and forgot to add it back, thanks, fixed!

File: target/linux/ramips/dts/R6120.dts (hunk 7 above)
Comment (Brother-Lal) [reply target/linux/ramips/dts/R6120.dts]: Wan has +1

**Comment by Brother-Lal** [comment]
Thanks for the docu-sheet!
I fixed the errors, and had a look at the existing dgn35000 firmware file.
Seems its version 0 and the netgear r6120 and r6220 have the version 1, with zipped rootfs.
But i could modify the tool to zip or not zip the rootfs depending on the version number.
I'll try some other things with the firmware generator, to get it to work with the webinterface, and update.
Feel free to point out any other things i can fix in the meantime :)

**Comment by blogic** [comment]
patch fails to apply, please rebase it

**Comment by Brother-Lal** [comment]
Patch is rebased, and adapted to new power led handling.
Current state: webinterface upload of firmware still fails, and the wifi leds stopped working.
I'll fix the wifi leds soonish, and then i hope we can merge, as flashing via nmrpflash is possible.

I do have the r6220 here also, as soon as i add this support to the firmware tool, i'll try to see and compare if i can flash via webinterface or what is wrong with the generated file. (In a separate PR)

File: target/linux/ramips/base-files/etc/board.d/01_leds (hunk 2 above)
Comment (blogic) [reply target/linux/ramips/base-files/etc/board.d/01_leds]: @Brother-Lal this needs to be fixed before we can merge it.

**Comment by Brother-Lal** [comment]
Updated and fixed wifi leds, switch declaration.
Should be mergeable now. @mkresin 

Webinterface flash is still not working, but will be worked on later.

**Comment by mkresin** [comment]
I fix the remaining issues of the PR and pushed the cleanup to [my staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary).

Either I broke something while doing the cleanup or there is a bug in the image code since the beginning. If I build an image, the following files are added to my local tree:

- target/linux/ramips/image/fs.zip
- target/linux/ramips/image/rootfs.img

What ever the reason for this behaviour is, it is wrong and need to be fixed.

**Comment by Brother-Lal** [comment]
Yes, this is a big, thise are temporary files that should be deleted after successful

**Comment by Brother-Lal** [comment]
Wrong button on mobile, sry...
Those are temp files that i forgot to delete in the image generation script.
I'll fix it this afternoon.

**Comment by mkresin** [comment]
Beside of that, shouldn't the files stored in a temporary directory? Possibly in the same directory like the input file?

**Comment by Brother-Lal** [comment]
Yes, they should be stored in a temp directory - I thought the image generation happens in such a directory, as the tempfile is just generated inside the current working dir.
I probably would then need to extract the current image generation path from the path for the sysupgrade image and prepend that...
Or would a `cd $(TMPDIR)` before the mksercommfw tool execution in the image makefile be an ok solution, if the files get erased after that?

**Comment by mkresin** [comment]
> Or would a cd $(TMPDIR) before the mksercommfw tool execution in the image makefile be an ok solution, if the files get erased after that?

No. Please have a look how the other fw-utils are doing it.

**Comment by Brother-Lal** [comment]
updated to remove the files and create them in the tmp directory in the first place.
also makes the mv @ in the image makefile unnecessary.

**Comment by mkresin** [comment]
Would have been easier for both of us if you had updated the PR with the code from my staging tree before applying more changes.

I updated the code in my [staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary) with the changes you have done. Hopefully I didn't missed anything.

Please build and runtime test an image build from [my staging tree](https://git.openwrt.org/?p=openwrt/staging/mkresin.git;a=summary) to ensure I imported all changes and everything still works as it should be.

**Review by mkresin — APPROVED** [review]

**Comment by Brother-Lal** [comment]
Works as intended! Thank you for persevering on that one ...

**Comment by mkresin** [comment]
Glad to hear. Added the commits to my push queue and will commit them to master later the day.

**Comment by adrianschmutzler** [comment]
@Brother-Lal Just stumbled about R6120 not having MAC+1 used. From earlier comments I found that this would be the WAN MAC. Is there a reason why this one is not set in 02_network? Otherwise, I would send a fix for that.

**Comment by Brother-Lal** [comment]
@adrianschmutzler Sadly i neither got my notes nor the hardware still here, so i can't really comment.
There were some additionall changes in between and somebody else integrated and fixed further on that hw.
But it can very realistic be an oversight on my part that i forgot to integrate that correctly.
Would be good to confirm that with original HW and fix if necessary. I will see if i can figure out who has access to one,

**Comment by adrianschmutzler** [comment]
@Brother-Lal Save your time. I've realized that there is a default case to set wan_mac to eth0 + 1. So, your device has had correct WAN MAC address from the beginning. Sorry for the confusion. (Although I dislike the idea of having a default case for stuff like that.)

**Comment by rotanid** [comment]
@blocktrron has the device :)

**Pull request closed without merging**